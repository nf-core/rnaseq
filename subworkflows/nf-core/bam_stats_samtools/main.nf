//
// Run SAMtools stats, flagstat and idxstats
//

include { SAMTOOLS_STATS    } from '../../../modules/nf-core/samtools/stats/main'
include { SAMTOOLS_IDXSTATS } from '../../../modules/nf-core/samtools/idxstats/main'
include { SAMTOOLS_FLAGSTAT } from '../../../modules/nf-core/samtools/flagstat/main'

workflow BAM_STATS_SAMTOOLS {
    take:
    ch_bam_bai // channel: [ val(meta), path(bam), path(bai) ]
    ch_fasta   // channel: [ val(meta), path(fasta) ]

    main:

    SAMTOOLS_STATS ( ch_bam_bai, ch_fasta )

    SAMTOOLS_FLAGSTAT ( ch_bam_bai )

    SAMTOOLS_IDXSTATS ( ch_bam_bai )

    emit:
    stats    = SAMTOOLS_STATS.out.stats       // channel: [ val(meta), path(stats) ]
    flagstat = SAMTOOLS_FLAGSTAT.out.flagstat // channel: [ val(meta), path(flagstat) ]
    idxstats = SAMTOOLS_IDXSTATS.out.idxstats // channel: [ val(meta), path(idxstats) ]

}
