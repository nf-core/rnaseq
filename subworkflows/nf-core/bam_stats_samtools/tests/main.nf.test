nextflow_workflow {

    name "Test Workflow BAM_STATS_SAMTOOLS"
    script "../main.nf"
    workflow "BAM_STATS_SAMTOOLS"
    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "bam_stats_samtools"
    tag "subworkflows/bam_stats_samtools"
    tag "samtools"
    tag "samtools/flagstat"
    tag "samtools/idxstats"
    tag "samtools/stats"

    test("test_bam_stats_samtools_single_end") {

        when {
            params {
                outdir   = "$outputDir"
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:true ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.single_end.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.single_end.sorted.bam.bai', checkIfExists: true)
                ])
                input[1] = Channel.of([
                    [ id:'genome' ],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out.stats).match("test_bam_stats_samtools_single_end_stats") },
                { assert snapshot(workflow.out.flagstat).match("test_bam_stats_samtools_single_end_flagstats") },
                { assert snapshot(workflow.out.idxstats).match("test_bam_stats_samtools_single_end_idxstats") }
            )
        }
    }

    test("test_bam_stats_samtools_paired_end") {

        when {
            params {
                outdir   = "$outputDir"
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:true ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                ])
                input[1] = Channel.of([
                    [ id:'genome' ],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out.stats).match("test_bam_stats_samtools_paired_end_stats") },
                { assert snapshot(workflow.out.flagstat).match("test_bam_stats_samtools_paired_end_flagstats") },
                { assert snapshot(workflow.out.idxstats).match("test_bam_stats_samtools_paired_end_idxstats") }
            )
        }
    }

    test("test_bam_stats_samtools_paired_end_cram") {

        when {
            params {
                outdir   = "$outputDir"
            }
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', single_end:false ], // meta map
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/cram/test.paired_end.sorted.cram', checkIfExists: true),
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/illumina/cram/test.paired_end.sorted.cram.crai', checkIfExists: true)
                ])
                input[1] = Channel.of([
                    [ id:'genome' ],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/genome/genome.fasta', checkIfExists: true)
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out.stats).match("test_bam_stats_samtools_paired_end_cram_stats") },
                { assert snapshot(workflow.out.flagstat).match("test_bam_stats_samtools_paired_end_cram_flagstats") },
                { assert snapshot(workflow.out.idxstats).match("test_bam_stats_samtools_paired_end_cram_idxstats") }
            )
        }
    }

}
