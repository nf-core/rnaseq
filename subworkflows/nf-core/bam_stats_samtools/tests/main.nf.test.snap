{
    "test_bam_stats_samtools_paired_end_cram_flagstats": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.flagstat:md5,a53f3d26e2e9851f7d528442bbfe9781"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2023-11-06T09:31:26.194017574"
    },
    "test_bam_stats_samtools_paired_end_stats": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.stats:md5,ddaf8f33fe9c1ebe9b06933213aec8ed"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2024-02-13T16:45:06.230091746"
    },
    "test_bam_stats_samtools_paired_end_flagstats": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.flagstat:md5,4f7ffd1e6a5e85524d443209ac97d783"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2024-01-18T17:17:27.717482"
    },
    "test_bam_stats_samtools_single_end_flagstats": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.flagstat:md5,2191911d72575a2358b08b1df64ccb53"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2023-11-06T09:26:10.340046381"
    },
    "test_bam_stats_samtools_paired_end_cram_idxstats": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.idxstats:md5,e179601fa7b8ebce81ac3765206f6c15"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2023-11-06T09:31:26.207052003"
    },
    "test_bam_stats_samtools_single_end_stats": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.stats:md5,dc178e1a4956043aba8abc83e203521b"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2024-02-13T16:44:57.442208382"
    },
    "test_bam_stats_samtools_paired_end_idxstats": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.idxstats:md5,df60a8c8d6621100d05178c93fb053a2"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2024-01-18T17:17:27.726719"
    },
    "test_bam_stats_samtools_single_end_idxstats": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": true
                    },
                    "test.idxstats:md5,613e048487662c694aa4a2f73ca96a20"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2023-11-06T09:26:10.349439801"
    },
    "test_bam_stats_samtools_paired_end_cram_stats": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test.stats:md5,d3345c4887f4a9ea4f7f56405b495db0"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.01.0"
        },
        "timestamp": "2024-02-13T16:45:14.997164209"
    }
}